### PR TITLE
Complete rewrite of JSON-RPC

### PR DESCRIPTION
**Pull request by jprupp**
A very different JSON-RPC interface based on type classes.

### CODE REVIEW

File: tests/data/invalid.json [hunk 1]
```diff
@@ -1,6 +0,0 @@
-{"jsonrpc": "2.0", "method": "foobar, "params": "bar", "baz]
-{"jsonrpc": "2.0", "method": 1, "params": "bar"}
-{"jsonrpc": "2.0", "method"
-[]
-[1,2,3]
-{"foo": "boo"}
```
Comment (m0dev76) [inline tests/data/invalid.json:6]: Why remove these tests? Are they no longer relevant?

Comment (jprupp) [inline tests/data/invalid.json:6]: Yes, these tests are no longer relevant.

Comment (m0dev76) [inline tests/data/invalid.json:6]: How so? They seem to be relevant for JSON-RPC. Do you have equivalent tests for these somewhere? Maybe I missed them

Comment (jprupp) [inline tests/data/invalid.json:6]: The way the library is written now makes these tests redundant. I can add them if you wish, but I don't see much value in them anymore.

**Comment by m0dev76** [comment]
Are they redundant because other tests check the same thing? Otherwise you usually want to keep tests because they give you guarantees that you don't break something while refactoring in the future. Removing tests because the library passes them doesn't make much sense :) Maybe you have a good reason to remove them, but your replies have been a bit vague.

**Comment by jprupp** [comment]
I used to have generic data types where I could accommodate a partially-parsed Aeson Value that was a structurally valid JSON-RPC request or response. It made sense to attempt to parse anything to these data types, including invalid data.

The current iteration of the code declares functions that take an element of type Value, and outputs data that belongs to user-defined data types inside the Aeson Parser monad. This removal of the intermediate data type makes the type system deal with invalid data. The parser just returns failure upon being unable to create Haskell data, it is the only thing that it can do.

That being said, the tests may still be performed, and I'm adding them back, because they may still be used to detect exceptions that cannot be handled by the Parser monad.

File: Network/Haskoin/Stratum/RPC.hs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+, encodeNotif
+, parseNotifParams
+, parseNotif
+  -- ** Encoding
+) where
+
+import Control.Applicative ((<$>))
+import Data.Aeson.Types
+import Data.Text (Text)
+import Data.IntMap.Strict (IntMap)
```
Comment (m0dev76) [inline Network/Haskoin/Stratum/RPC.hs]:
```
Network/Haskoin/Stratum/RPC.hs:31:1: Warning:
    The import of `Data.IntMap.Strict' is redundant
      except perhaps to import instances from `Data.IntMap.Strict'
    To import instances alone, use: import Data.IntMap.Strict()
```

Comment (jprupp) [inline Network/Haskoin/Stratum/RPC.hs]: Yup, noticed it. I will update commit soon.

**Comment by m0dev76** [comment]
ACK

**Pull request merged**